### ISSUE TITLE
mAP and r1 of hnm

### ISSUE THREAD

**Issue by MOONBEAR0703** [ergysr/DeepCC#49]
Hi, it is strange.  After I added augmentation, mAP and r1 increased, but after I trained hnm, mAP and r1 decreased, and even lower than the fundamental AWTL.

**Comment by ergysr** [comment]
Did this happen once, or repeatedly? And when you say augment, is it flip&crop, or the imgaug augmentation?

**Comment by MOONBEAR0703** [comment]
In fact, I have tried the hnm only twice, but the results are the same, mAP=0.488,r1=0.693. 
For the first time, it was the hnm without any augmentation; for the second time, it was the hnm with the imgaug augmentation. (for #12 issues, several days ago I updated my tensorflow to 1.10.0 and it worked well)

**Comment by ergysr** [comment]
These scores seem pretty low compared to what you should be getting. Are you also getting low scores for HTRI + hnm?

**Comment by MOONBEAR0703** [comment]
What does HTRI represent for? 
AWTL+imgaug augmentation: mAP=0.559480, r1=0.742819
AWTL: mAP=0.547712, r1=0.740921

**Comment by ergysr** [comment]
I ran `train_duke_hnm` (AWTL + augmentation + hnm)  overnight and it is producing scores similar to the ones I shared. So not sure why you are getting low scores.

**Comment by MOONBEAR0703** [comment]
I guess there must be some parameters wrong.
The results of hnm are always the same, although I have changed `net.loss = 'batch_hard'`and `net.augment = true`.
 mAP = 0.488218, r1 precision = 0.692549

**Comment by MOONBEAR0703** [comment]
I deleted all folders under the src/triplet_reid/experiment.
Then I retrained duke_hnm and mAP=0.563949,r1 precision =0.755835.

**Comment by MOONBEAR0703** [comment]
Hi, what if I want to get my `DukeMTMC MultiCam Results` with the (AWTL + augmentation + hnm ) reid features gained before? I changed `opts.net.experiment_root = 'experiments/demo_hnm'` in which it stored the results of (AWTL  + augmention + hnm), `demo_hnm` replaced `demo_weighted_triplet`.
But the results are the same as the first time I trained.
********************* Your DukeMTMCT MultiCam Results *********************
IDF1   IDP    IDR
79.75  79.88  79.63
Also, in the terminal window, it said:

> [h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6e3e7e1ec0] Missing reference picture, default is 0
[h264 @ 0x7f6e3e7e1ec0] decode_slice_header error
[h264 @ 0x7f6e3e7e1ec0] Missing reference picture, default is 0
[h264 @ 0x7f6e3e7e1ec0] decode_slice_header error
[h264 @ 0x7f6e3e7e1ec0] Missing reference picture, default is 0
[h264 @ 0x7f6e3e7e1ec0] decode_slice_header error
[h264 @ 0x7f6e3e7e1ec0] Missing reference picture, default is 0
[h264 @ 0x7f6e3e7e1ec0] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error
[h264 @ 0x7f6fc7af8040] Missing reference picture, default is 0
[h264 @ 0x7f6fc7af8040] decode_slice_header error

**Comment by ergysr** [comment]
Fig 6 in the paper illustrates that similar IDF1 performance can be obtained when the rank-1 accuracy is above a certain level. This means that one doesn't need the best appearance features to track well, so whether you use HNM or not it will not make a different in tracking, only in rank-1 accuracy.

The messages in the terminal window are warnings from the video decoder and you can safely ignore them.

**Comment by vedantbhatia** [comment]
> Did this happen once, or repeatedly? And when you say augment, is it flip&crop, or the imgaug augmentation?


Going to refer to 
` mkdir('src/triplet-reid/experiments/')
opts = get_opts();
train_duke(opts);
embed(opts);
evaluation_res_duke_fast(opts);`
 as [1].

To achieve your results on the Duke dataset, do we have to implement the augmentation (mentioned in the paper) through imgaug on our own?
Per my understanding;
1. AWTL - Just run [1] without installing imgaug
2. AWTL ( + Aug ) - run [1] after installing imgaug and modifying images folder in get_opts.m
3. AWTL (+ Aug + hnm) - same as 2. but with 'train_duke_hnm;' instead of train_duke(opts); in [1]

Am I correct?

**Comment by vedantbhatia** [comment]
> For bullet point 1, you need to set opts.net.augment = False before running [1]. This will default to the simple flip&crop augmentation.

Thanks.

Is the code to replicate the AWTL (2-stream) re-id results available as well?

**Comment by ergysr** [comment]
I didn't include that code in the repository. To get something like the 2-stream effect you can train the network twice. Then the image feature is the concatenation from the ensemble of two network outputs (128d + 128d).

**Comment by vedantbhatia** [comment]
> I didn't include that code in the repository. To get something like the 2-stream effect you can train the network twice. Then the image feature is the concatenation from the ensemble of two network outputs (128d + 128d).

Thanks. Could you please give me an idea of the sort of changes I would have to make to the code?

**Comment by ergysr** [comment]
For ReID evaluation you can modify [here ](https://github.com/ergysr/DeepCC/blob/ab89f0a82a6b603d87c508b2c3ddc217d7075d6e/src/external/DukeMTMC-reID_evaluation/evaluation_res_duke_fast.m#L10-L16). Instead of reading a single file for query and gallery image features, you can read two output files and concatenate their features. This is the quickest way, not necessarily the best.

For multi-camera association the features are loaded [here](https://github.com/ergysr/DeepCC/blob/ab89f0a82a6b603d87c508b2c3ddc217d7075d6e/src/L3-identities/getTrajectoryFeatures.m#L63).

**Comment by vedantbhatia** [comment]
> For ReID evaluation you can modify [here ](https://github.com/ergysr/DeepCC/blob/ab89f0a82a6b603d87c508b2c3ddc217d7075d6e/src/external/DukeMTMC-reID_evaluation/evaluation_res_duke_fast.m#L10-L16). Instead of reading a single file for query and gallery image features, you can read two output files and concatenate their features. This is the quickest way, not necessarily the best.
> 
> For multi-camera association the features are loaded [here](https://github.com/ergysr/DeepCC/blob/ab89f0a82a6b603d87c508b2c3ddc217d7075d6e/src/L3-identities/getTrajectoryFeatures.m#L63).

Thank you very much!
